### PR TITLE
DC-337: Video Thumbnail Inline Play Functionality

### PR DESCRIPTION
**Pull request by colbytcook**
## Jira

https://pegadigitalit.atlassian.net/browse/DC-337

## Summary

* Modified the Video Thumbnail component to handle both inline and modal video support.
* Adds a subtitle icon indicator.

## Details

* Refactored the Video Thumbnail schema to handle the new inline functionality option.
* Added `subtitles` and `video` props to  the schema.
* Added a inline video doc page.
* Updated all instances of the Video Thumbnail Component to include a `video.video_title` prop in the doc site.
* [WORK IN PROGRESS] Added Jest tests to the Video Thumbnail Component.

## How to test

* Review the code
* Pull down the branch
* Review the Video Thumbnail docs (a lot has changed within this section)

## Release notes

* Video Thumbnail Component now has a inline video option.
* Added a `subtitles` prop to the Video Thumbnail components. This prop will render a subtitles icon with a "Subtitles available" tooltip to a video thumbnail component.
* Added a `video` prop to the Video Thumbnail components. This prop will give a content author the ability to play videos inline.

### CODE REVIEW

File: packages/components/bolt-video-thumbnail/src/video-thumbnail.twig [hunk 1]
```diff
@@ -32,9 +32,28 @@
...
+        <div class="c-bolt-video-thumbnail__subtitles">
+          {% set icon %}
+            {% include '@bolt-elements-icon/icon.twig' with {
+              name: 'pega-subtitle',
+              size: 'small'
+            } only %}
+          {% endset %}
+          {% include '@bolt-components-tooltip/tooltip.twig' with {
+            trigger: icon,
+            content: 'Subtitles'
```
Comment (remydenton) [inline packages/components/bolt-video-thumbnail/src/video-thumbnail.twig:52]: Don't forget to add a |t here: `'Subtitles'|t`

File: packages/components/bolt-video-thumbnail/video-thumbnail.schema.js [hunk 2]
```diff
@@ -30,6 +30,12 @@ module.exports = {
       description:
         'Display the video duration that overlays the video thumbnail. This should only be plain text.',
     },
+    subtitles: {
```
Comment (remydenton) [inline packages/components/bolt-video-thumbnail/video-thumbnail.schema.js:33]: Naming nitpick, change the naming of `subtitles` to `has_subtitles`.

This isn't a convention we've followed elsewhere, but honestly we should.  It's general programming naming convention for booleans (see, for example, https://www.serendipidata.com/posts/naming-guidelines-for-boolean-variables).  "Subtitles" is especially confusing here because it sounds like maybe you should be passing the subtitles themselves.

**Review by mikemai2awesome — CHANGES_REQUESTED** [review]
@colbytcook As we discussed in a call, let's update the inline video player version.

1. Reduce to just one focusable area which is the entire thumbnail. Tooltip for subtitles icon will appear on hover and focus state of this area. This area needs a visible focus outline (`outline: var(--bolt-focus-ring);`) and to reset button styles there is a mixin (`bolt-button-native-styles-reset`).
2. Make sure the entire thumbnail is clickable, any UI inside (subtitles icon, duration, etc.) should not block it.

File: packages/components/bolt-video-thumbnail/src/video-thumbnail.scss [hunk 3]
```diff
@@ -70,6 +70,26 @@
   }
 }
 
+.c-bolt-video-thumbnail__content__inline {
```
Comment (mikemai2awesome) [inline packages/components/bolt-video-thumbnail/src/video-thumbnail.scss:73]: Let's just update this to `.c-bolt-video-thumbnail__content-video`. Set `display: none` so screen reader can't reach the hidden video embed until the button is clicked to reveal it.

File: packages/components/bolt-video-thumbnail/src/video-thumbnail.scss [hunk 4]
```diff
@@ -70,6 +70,26 @@
...
+  position: absolute;
+  top: 0;
+  right: 0;
+  bottom: 0;
+  left: 0;
+}
+
+// These updates are to make sure the button doesn't flash due to the image not loading immediately
+.js-bolt-video-thumbnail__inline-play {
```
Comment (mikemai2awesome) [inline packages/components/bolt-video-thumbnail/src/video-thumbnail.scss:82]: Cannot style JS classes, this would be `.c-bolt-video-thumbnail__inline-play-button`. To target it with JS, that would be a separate JS class.

File: packages/components/bolt-video-thumbnail/src/video-thumbnail.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    });
+  }
+
+  destroyElement(el) {
+    el.remove();
+  }
+
+  playVideo(el) {
+    el.currentTarget
+      .closest('.js-bolt-video-thumbnail__inline-play__target')
```
Comment (mikemai2awesome) [inline packages/components/bolt-video-thumbnail/src/video-thumbnail.js:37]: No BEM for JS classes. Just dash case.

File: packages/components/bolt-video-thumbnail/src/video-thumbnail.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
+export class BoltVideoThumbnail {
+  constructor(el) {
+    if (!el) return;
+    this.el = el;
+    this.init();
+  }
+
+  init() {
+    const inlineVideoPlayButton = this.el.querySelector(
+      '.js-bolt-video-thumbnail__inline-play',
```
Comment (mikemai2awesome) [inline packages/components/bolt-video-thumbnail/src/video-thumbnail.js:10]: No BEM for JS classes. Just dash case.

File: packages/components/bolt-video-thumbnail/src/video-thumbnail.twig [hunk 7]
```diff
@@ -12,15 +12,33 @@
...
 ] %}
 
+{% if has_subtitles %}
+  {% set subtitles_label = 'Subtitles available'|t %}
+{% endif %}
+
 {# Template #}
 <div {{ attributes.addClass(classes) }}>
-  <div class="c-bolt-video-thumbnail__content">
+  <div class="c-bolt-video-thumbnail__content js-bolt-video-thumbnail__inline-play__target">
```
Comment (mikemai2awesome) [inline packages/components/bolt-video-thumbnail/src/video-thumbnail.twig:24]: No BEM for JS classes. Just dash case.

**Review by mikemai2awesome — CHANGES_REQUESTED** [review]

**Review by mikemai2awesome — CHANGES_REQUESTED** [review]

File: packages/components/bolt-video-thumbnail/src/video-thumbnail.twig (hunk 1 above)
Comment (colbytcook) [reply packages/components/bolt-video-thumbnail/src/video-thumbnail.twig]: @remydenton updated

File: packages/components/bolt-video-thumbnail/src/video-thumbnail.twig [hunk 8]
```diff
@@ -12,15 +12,29 @@
...
 {# Template #}
 <div {{ attributes.addClass(classes) }}>
   <div class="c-bolt-video-thumbnail__content">
+    {% if video %}
+      <div class="c-bolt-video-thumbnail__content__inline">
+        {{ video }}
+      </div>
+      {% include '@bolt-elements-button/button.twig' with {
+        display: 'block',
+        content: 'Play the video',
```
[the remark this answers is not in the document]
Comment (colbytcook) [reply packages/components/bolt-video-thumbnail/src/video-thumbnail.twig]: @remydenton updated

File: packages/components/bolt-video-thumbnail/src/video-thumbnail.js [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    });
+  }
+
+  destroyElement(el) {
+    el.remove();
+  }
+
+  playVideo(el) {
+    el.currentTarget.previousElementSibling
+      .querySelector('.c-base-video')
```
[the remark this answers is not in the document]
Comment (colbytcook) [reply packages/components/bolt-video-thumbnail/src/video-thumbnail.js]: Updated

File: packages/components/bolt-video-thumbnail/src/video-thumbnail.scss [hunk 10]
```diff
@@ -70,6 +70,26 @@
...
+  &:before,
+  &:after {
+    background-color: var(--bolt-color-white);
+  }
+
+  &:before {
+    opacity: 1;
+  }
+}
```
[the remark this answers is not in the document]
Comment (colbytcook) [reply packages/components/bolt-video-thumbnail/src/video-thumbnail.scss:99]: @remydenton updated

File: packages/components/bolt-video-thumbnail/video-thumbnail.schema.js (hunk 2 above)
Comment (colbytcook) [reply packages/components/bolt-video-thumbnail/video-thumbnail.schema.js]: @remydenton updated

File: packages/components/bolt-video-thumbnail/video-thumbnail.schema.js [hunk 11]
```diff
@@ -25,20 +25,48 @@ module.exports = {
...
-    duration: {
+    chip: {
       type: 'any',
       description:
-        'Display the video duration that overlays the video thumbnail. This should only be plain text.',
+        'Display a chip (or chip-list) that overlays the video thumbnail. Positioned before the video title. Chip or Chip-list component is expected here.',
     },
-    video_title: {
-      type: 'any',
+    video_title_display: {
```
[the remark this answers is not in the document]
Comment (colbytcook) [reply packages/components/bolt-video-thumbnail/video-thumbnail.schema.js:33]: @mikemai2awesome due to the accessibility needs the video.title is required (the title can be automatically fed into the component from Drupal). The video_title_display is different in that it is an optional display that adds a title chip, you see this behavior [here](https://boltdesignsystem.com/pattern-lab/?p=components-video-thumbnail-title-and-chips). The video.title will be fed into the title chip and the aria-label for the button automatically. 

So to summarize, the title (video.title) is required and will always be passed in but the title chip (video_title_display) is an optional visual display type that needs separate prop control for content authors. Let me know if you have anymore questions/suggestions on this.

File: packages/components/bolt-video-thumbnail/video-thumbnail.schema.js [hunk 12]
```diff
@@ -25,20 +25,48 @@ module.exports = {
...
+          description:
+            'Display the video duration that overlays the video thumbnail. This should only be plain text.',
+        },
+        has_subtitles: {
+          type: 'boolean',
+          description:
+            'Display a subtitles icon that overlays the video thumbnail.',
+          default: false,
+        },
+        video_title: {
```
[the remark this answers is not in the document]
Comment (colbytcook) [reply packages/components/bolt-video-thumbnail/video-thumbnail.schema.js]: @mikemai2awesome updated

File: packages/components/bolt-video-thumbnail/src/video-thumbnail.twig [hunk 13]
```diff
@@ -7,34 +7,75 @@
...
 <div {{ attributes.addClass(classes) }}>
+  {% if video.content or button_attributes %}
+    <button {{ button_attributes.addClass('c-bolt-video-thumbnail__inline-button')|without('aria-label')|without('type') }} type="button" aria-label="{{ play_video_label|t }}">
+  {% endif %}
   <div class="c-bolt-video-thumbnail__content">
     {{ content }}
   </div>
-  {% if chip or video_title %}
-    <div class="c-bolt-video-thumbnail__meta">
+  {% if chip or video_title_display %}
```
[the remark this answers is not in the document]
Comment (colbytcook) [reply packages/components/bolt-video-thumbnail/src/video-thumbnail.twig:39]: @mikemai2awesome see comment above

File: packages/components/bolt-video-thumbnail/video-thumbnail.schema.js [hunk 14]
```diff
@@ -25,20 +25,48 @@ module.exports = {
...
+        'Display a chip (or chip-list) that overlays the video thumbnail. Positioned before the video title. Chip or Chip-list component is expected here.',
     },
-    video_title: {
-      type: 'any',
+    video_title_display: {
+      type: 'boolean',
       description:
-        'Display the video title that overlays the video thumbnail. This should only be plain text.',
+        'Display a video title chip that overlays the video thumbnail.',
+      default: false,
```
Comment (remydenton) [inline packages/components/bolt-video-thumbnail/video-thumbnail.schema.js:37]: I think it's the right thing to do to have this be hidden by default.

Currently, there are asset injectors that hide this title with CSS where it's not wanted, e.g. https://www.pega.com/admin/config/development/asset-injector/css/suppress_video_meta_title.  But I believe that is that majority of the time, so I think it's better to hide it by default, and add an option to the UI to display it in the (relatively rare) cases that's desired.

File: packages/components/bolt-video-thumbnail/src/video-thumbnail.twig [hunk 15]
```diff
@@ -7,34 +7,75 @@
...
+  {% set play_video_label = 'Play the video' %}
+  {% set _has_subtitles = video.has_subtitles ? ', this video has subtitles available.' : '' %}
+  {% set _title = video.title ? ' ' ~ video.title : '' %}
+  {% set play_video_label = play_video_label ~ _title ~ _has_subtitles %}
+{% endif %}
+
 {# Template #}
 <div {{ attributes.addClass(classes) }}>
+  {% if video.content or button_attributes %}
+    <button {{ button_attributes.addClass('c-bolt-video-thumbnail__inline-button')|without('aria-label')|without('type') }} type="button" aria-label="{{ play_video_label|t }}">
```
Comment (remydenton) [inline packages/components/bolt-video-thumbnail/src/video-thumbnail.twig:34]: `|t `doesn't work with a variable; you should instead use it on a literal string.  More info at https://www.drupal.org/docs/7/api/localization-api/localization-api-overview#s-introduction-to-interface-translation-in-drupal

In short, t should be used earlier on when you first declare the string:
```
{% set play_video_label = 'Play the video'|t %}
```

**Review by mikemai2awesome — APPROVED** [review]

**Review by danielamorse — APPROVED** [review]
@colbytcook updated the JS as we discussed.

**Review by remydenton — APPROVED** [review]

**Review by mikemai2awesome — APPROVED** [review]
